pkg/kv/kvserver/replica_write.go

```diff
@@ -884,3 +889,108 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
 	ms.Reset()
 	mvccStatsPool.Put(ms)
 }
+
+// overlapsUnreplicatedRangeIDLocalKeys checks if the provided span overlaps
+// with any unreplicated rangeID local keys.
+func overlapsUnreplicatedRangeIDLocalKeys(span roachpb.Span) error {
+	// getRangeIDLocalSpan is a helper function that returns the full
+	// unreplicated RangeID span for a given RangeID.
+	getRangeIDLocalSpan := func(rangeID roachpb.RangeID) roachpb.Span {
+		return roachpb.Span{Key: keys.MakeRangeIDUnreplicatedPrefix(rangeID),
+			EndKey: keys.MakeRangeIDUnreplicatedPrefix(rangeID).PrefixEnd()}
+	}
+
+	fullRangeIDLocalSpans := roachpb.Span{
+		Key:    keys.MakeRangeIDUnreplicatedPrefix(0),
+		EndKey: keys.LocalRangePrefix,
+	}
+
+	// If the provided span is completely outside the rangeID local spans for any
+	// rangeID, then there is no overlap with any rangeID local keys.
+	if !spanset.Overlaps(fullRangeIDLocalSpans, span) {
+		return nil
+	}
+
+	// Overlapping the full rangeID local span. We must be overlapping rangeID
+	// local keys.
+	if spanset.Contains(span, fullRangeIDLocalSpans) {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
```

Comment:
`Contains` can return true if `A == B`, right? But we're specifically interested whether there are parts of `A` that are inside `B`, as well as outside.